File: Projects/JavaProject/src/com/Basics_of_Java/TypeCasting.java
package com.Basics_of_Java;

public class TypeCasting {
    public static void main(String[] args) {
        /*
         * Implicit Type Casting (Widening Casting)
         * Happens automatically when converting a smaller data type into a larger one
         * (e.g., int to double).
         * 
         * byte → short → int → long → float → double
         * 
         */
        int num = 10;
        double doubleNum = num; // Implicit casting (Widening)

        System.out.println("Integer: " + num);
        System.out.println("Converted to Double: " + doubleNum);

        /*
         * Explicit Type Casting (Narrowing Casting)
         * Requires manual conversion when converting a larger data type into a smaller one
         * (e.g., double to int).
         * 
         * double → float → long → int → short → byte
         * 
         */

        double doubleNum1 = 10.75;
        int num1 = (int) doubleNum1; // Explicit casting (Narrowing)

        System.out.println("Double: " + doubleNum1);
        System.out.println("Converted to Integer: " + num1);

    }
}

File: Projects/PythonProject/Basics_of_Python/TypeCasting.py
def main():
    print("=== Implicit Type Casting (Widening) ===")
    num = 10
    double_num = float(num)  # Python automatically handles this
    print("Integer:", num)
    print("Converted to Float (Double-like):", double_num)

    print("\n=== Explicit Type Casting (Narrowing) ===")
    double_num1 = 10.75
    num1 = int(double_num1)  # Manual conversion, decimal part is truncated
    print("Float (Double-like):", double_num1)
    print("Converted to Integer:", num1)

    print("\n=== String to Integer Conversion ===")
    s1 = "123"
    try:
        num2 = int(s1)
        print("Converted integer (s1):", num2)
    except ValueError:
        print("s1 is not a valid integer.")

    s2 = "123abc"
    try:
        num3 = int(s2)
        print("Converted integer (s2):", num3)
    except ValueError:
        print("s2 is not a valid integer.")

    s3 = "12.34"
    try:
        num4 = int(float(s3))  # Convert to float first, then to int
        print("Converted from decimal string (s3):", num4)
    except ValueError:
        print("s3 is not a valid number.")

    s4 = "₹100"
    try:
        cleaned_s4 = ''.join(filter(str.isdigit, s4))
        num5 = int(cleaned_s4)
        print("Converted from currency string (s4):", num5)
    except ValueError:
        print("s4 is not a valid number.")

if __name__ == "__main__":
    main()
